Modules/posixmodule.c

```diff
@@ -10073,8 +10073,27 @@ os__getdiskusage_impl(PyObject *module, path_t *path)
     Py_BEGIN_ALLOW_THREADS
     retval = GetDiskFreeSpaceExW(path->wide, &_, &total, &free);
     Py_END_ALLOW_THREADS
-    if (retval == 0)
+    if (retval == 0) {
+        if (GetLastError() == ERROR_DIRECTORY) {
+            PyObject *result = NULL;
+            wchar_t *wpath = NULL;
+            path_t dir_path = *path;
+
+            wpath = PyMem_New(wchar_t, dir_path.length + 1);
+            if (wpath == NULL)
+                return PyErr_NoMemory();
+
+            wcscpy_s(wpath, dir_path.length + 1, path->wide);
+            _dirnameW(wpath);
+            dir_path.wide = wpath;
+
+            result = os__getdiskusage_impl(module, &dir_path);
```

Comment:
I'd call `GetDiskFreeSpaceExW` inline on the parent directory and add a `goto success` statement. I dont see the need for a loop or recursion.